syntax = "proto3";
package todos;

message User{
    int32 id = 1;
    string userName = 2;
    string email = 3;
}

message Comment{
    int32 id = 1;
    string comment = 2 ;
    User user = 3;
}

message CreateTodoRequest{
    int32 user_id = 1;
    string title = 2;
    string description = 3;
}

message CreateTodoResponse{
    string message = 1;
}

message UpdateTodoRequest{
    int32 user_id =1;
    int32 todo_id = 2;
    string title = 3;
    string description = 4;
}

message UpdateTodoResponse{
    string message = 1;
}

message DeleteTodoRequest{
    int32 todo_id = 1;
    int32 user_id = 2;
}

message DeleteTodoResponse{
    string message = 1;
}

message CreateCommentRequest{
    int32 user_id = 1;
    int32 todo_id = 2;
    string comment = 3;
}

message CreateCommentResponse{
    string message = 1;
}

message GetTodoRequest{
    int32 user_id = 1;
    optional string title = 2;
}

message GetTodos{
    int32 id = 1;
    string title = 2;
    string description = 3;
    User user =4;
    repeated Comment comments = 5;
}

message GetTodoResponse{
   repeated GetTodos todos  = 1;
}

message GetTodoByIdRequest{
    int32 todo_id = 1;
    int32 user_id =2;
}

message GetCommentsRequest{
    int32 todo_id = 1;
}

message GetCommentsResponse{
    repeated Comment comments = 1;
}

service TodoService{
    rpc CreateTodo (CreateTodoRequest) returns(CreateTodoResponse){}
    rpc GetTodo (GetTodoRequest) returns (GetTodoResponse) {}
    rpc DeleteTodo (DeleteTodoRequest) returns (DeleteTodoResponse) {}
    rpc UpdateTodo (UpdateTodoRequest) returns (UpdateTodoResponse) {}
    rpc GetTodoById (GetTodoByIdRequest) returns (GetTodoResponse) {}
    rpc CreateComment (CreateCommentRequest) returns (CreateCommentResponse) {}
    rpc GetComments (GetCommentsRequest) returns (GetCommentsResponse) {}
}
